### PR TITLE
Improve process of creating new courses and units

### PR DESCRIPTION
**Pull request by dmcavoy**
Create a new flow for creating UnitGroups and Scripts. Force people to set family_name and version_year if the object is acting as a course. Keep the slugs consistent with family_name and version_year if an objects is acting a course.

Follow up work:
- Do not allow editing of family_name and version_year once they have been set

### New Course

https://user-images.githubusercontent.com/208083/145111091-b8d373a5-86af-4f5f-b733-506bafcb8e00.mov

### New Unit in Course

https://user-images.githubusercontent.com/208083/145111088-1cd3257c-bfcb-4c2b-b625-ee633a0d368f.mov

### New Unit as Course

https://user-images.githubusercontent.com/208083/145111066-18e07622-a6d9-45fc-82c1-611c92e9c064.mov

## Links

[Jira](https://codedotorg.atlassian.net/browse/PLAT-1436)

## Testing story

- Added a bunch of new unit tests
- Manually added a unit and course

### CODE REVIEW

File: dashboard/app/views/courses/new.html.haml [hunk 1]
```diff
@@ -1,3 +1,8 @@
+%h1= "New Course"
+%script{src: webpack_asset_path('js/courses/new.js'), data: {families: UnitGroup.family_names.to_json, versionYearOptions: UnitGroup.get_version_year_options.to_json}}
```
Comment (davidsbailey) [inline dashboard/app/views/courses/new.html.haml:2]: thank you for bringing this into React!

**Comment by davidsbailey** [comment]
> Follow up work:
> 
> * Do not allow editing of family_name and version_year once they have been set

when you get to this, perhaps "is a standalone unit" should not be editable either?

File: apps/src/lib/levelbuilder/NewCourseFields.jsx [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+          onChange={e => {
+            setSelectedFamilyName(e.target.value);
+            props.setFamilyName(e.target.value);
+          }}
+          disabled={newFamilyName !== ''}
+        >
+          <option value={''}>(None)</option>
+          {props.families.map(familyOption => (
+            <option key={familyOption} value={familyOption}>
+              {familyOption}
```
Comment (davidsbailey) [inline apps/src/lib/levelbuilder/NewCourseFields.jsx:27]: would it make sense to exclude family names where the existing unit in that family is "unversioned"? I'm not sure exactly what the right logic would be, but it seems like we could remove a lot of options from this list which the user would / should never pick.

File: apps/src/lib/levelbuilder/NewCourseFields.jsx [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        </select>
+        <span>
+          or{' '}
+          <input
+            type="text"
+            value={newFamilyName}
+            className="familyNameInput"
+            style={styles.smallInput}
+            onChange={e => {
+              setNewFamilyName(e.target.value);
```
Comment (davidsbailey) [inline apps/src/lib/levelbuilder/NewCourseFields.jsx:39]: should/could we disallow family names which are already taken when the user types one in here? I would assume they are expecting to get a new family name when they type in their own answer.

File: apps/src/lib/levelbuilder/course-editor/NewCourseForm.jsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        versionYearOptions={props.versionYearOptions}
+        familyName={familyName}
+        setFamilyName={setFamilyName}
+        versionYear={versionYear}
+        setVersionYear={setVersionYear}
+      />
+      {familyName !== '' && versionYear !== '' && (
+        <div>
+          <label>
+            The Course Slug for this course will be:
```
Comment (davidsbailey) [inline apps/src/lib/levelbuilder/course-editor/NewCourseForm.jsx:33]: I love that the editor will not have to pick the slug!

**Comment by dmcavoy** [comment]
> > Follow up work:
> > 
> > * Do not allow editing of family_name and version_year once they have been set
> 
> when you get to this, perhaps "is a standalone unit" should not be editable either?

The PR for it is here: https://github.com/code-dot-org/code-dot-org/pull/43984 and you are right it does make is standalone unit un-editable as well

File: apps/src/lib/levelbuilder/NewCourseFields.jsx (hunk 3 above)
Comment (davidsbailey) [reply apps/src/lib/levelbuilder/NewCourseFields.jsx:39]: also, should / could we reject entries that end in `-DDDD`?

**Review by davidsbailey — APPROVED** [review]
Awesome work Dani! I could go either way on the questions inline -- this will be such a big win once these fields are disabled on the course/unit edit pages.

File: apps/src/lib/levelbuilder/NewCourseFields.jsx (hunk 3 above)
Comment (dmcavoy) [reply apps/src/lib/levelbuilder/NewCourseFields.jsx:39]: I like this idea. I'm going to track it and follow up with another PR. https://codedotorg.atlassian.net/browse/PLAT-1514

File: apps/src/lib/levelbuilder/NewCourseFields.jsx (hunk 2 above)
Comment (dmcavoy) [reply apps/src/lib/levelbuilder/NewCourseFields.jsx:27]: I like this idea. I'm going to track it and follow up with another PR. https://codedotorg.atlassian.net/browse/PLAT-1514

**Pull request merged**